waspc/data/Generator/templates/react-app/src/actions/index.js

```diff
@@ -0,0 +1,74 @@
+import {
+  useMutation,
+  useQueryClient,
+} from 'react-query'
+
+export { configureQueryClient } from '../queryClient'
+
+export function useAction(actionFn, actionOptions) {
+  const queryClient = useQueryClient();
+
+  // TODO(filip): Do we allow all react query options or should we consider a list of allowed options?
+  let options = {}
+
+  if (actionOptions?.optimisticUpdates) {
+    const optimisticUpdatesConfig = makeOptimisticUpdatesConfig(actionOptions.optimisticUpdates)
+    options = makeOptimisticUpdateOptions(queryClient, optimisticUpdatesConfig)
+  }
+
+  return useMutation(actionFn, options)
+}
+
+
+function makeOptimisticUpdatesConfig(optimisticUpdatesConfig) {
+  return optimisticUpdatesConfig.map(({ getQuery, ...rest }) => ({
+    getQuery: (item) => parseQueryKey(getQuery(item)),
+    ...rest,
+  }))
+}
+
+function makeOptimisticUpdateOptions(queryClient, optimisticUpdatesConfig) {
+  async function onMutate(item) {
+    const queriesToUpdate = optimisticUpdatesConfig.map(({ getQuery, ...rest }) => ({
+      queryKey: getQuery(item),
+      ...rest,
+    }))
+
+    const queryCancellations = queriesToUpdate.map(
+      ({ query }) => queryClient.cancelQueries(query)
+    )
+
+    // Theoretically, we can be a bit faster. Instead of awaiting the
+    // cancellation of all queries, we could cancel and update them in parallel.
+    // However, awaiting cancellation probably doesn't take too much time.
+    await Promise.all(queryCancellations)
+
+    // We're using a Map to to correctly serialize query keys that contain objects
+    const previousData = new Map()
+    queriesToUpdate.forEach(({ queryKey, updateQuery }) => {
+      const previousDataForQuery = queryClient.getQueryData(queryKey)
+      const updateFn = (old) => updateQuery(item, old)
+      queryClient.setQueryData(queryKey, updateFn)
+      previousData.set(queryKey, previousDataForQuery)
+    })
+
+    return previousData
+  }
+
+  function onError(err, item, context) {
+    context.previousData.forEach(async (data, queryKey) => {
+      await queryClient.cancelQueries(queryKey)
+      queryClient.setQueryData(queryKey, data)
+    })
+  }
```

Comment:
TODO (filip): test this, couldn't get it to make sense. What errors is it supposed to catch? It triggers if I throw an exception inside onMutate and then complains that context.previousData is `undefined`.

Reply:
@shayneczyzewski said:

----

That does raise a good question, how do errors play a role in the counter approach? Are they automatically caught and then it just decrements it still?

Reply:
We've since abandoned the counter approach, so luckily no need to answer that :sweat_smile: (but yes, it would cause problems).

As for my comments, I found and fixed the issue, it was an error on my side.